Implements struct_lit_force_multiline

cburgdorf opened the pull request:
With struct_lit_force_multiline set to true
rustfmt won't ever mangle multiple struct properties
into one line.

Fixes #253

cburgdorf commented:
Not sure if it's what we ultimately want (see: https://github.com/nrc/rustfmt/issues/253#issuecomment-137873320 for more context) but it does the job for now ;)

nrc commented on src/config.rs:
```diff
@@ -122,6 +122,7 @@ create_config! {
     struct_trailing_comma: SeparatorTactic,
     struct_lit_trailing_comma: SeparatorTactic,
     struct_lit_style: StructLitStyle,
+    struct_lit_force_multiline: bool,
```
Could you make an enum for this rather than using a bool please? Otherwise in the future when we add more options then people who use this will have their configs broken.

nrc commented on src/expr.rs:
```diff
@@ -997,7 +997,11 @@ fn rewrite_struct_lit<'a>(context: &RewriteContext,
+            ListTactic::Vertical
+        } else {
+            ListTactic::HorizontalVertical
+        },
```
Once you've added the enum, could you move this to a method on that enum?

nrc commented:
Thanks for the PR. The strategy looks good, just a couple of refactoring nits. This also needs a test please.

cburgdorf commented:
I think I addressed your comments regarding the refactoring nits. But I think I need a little hand holding for the tests. I couldn't find a test for the struct literals (also I found the struct literals that are probably meant to be used for the tests)

nrc commented:
@cburgdorf so, tests: there are two kinds of test - source->target and target->target. There are appropriately named directories in the tests directory. For the source->target tests we test each source file, look for a similarly named target file and check that rustfmt(source) == target. Target->target tests are the same but just check for idempotence on all target files.

For struct lits, see struct_lits.rs and struct_lits_visual.rs. The latter is more instructive because the first line (a significant comment) change a value in the config under which the test is run.

So to make a test for this, you'll need to create new files in source and target with unformatted and formatted source code respectively. You'll need a comment line to set the new config option. You'll want a few different cases of struct literal (named vs unnamed fields, comments in different places, different numbers of fields (including zero)). Note that the tests don't have to be valid Rust code, they just have to parse.

Let me know if you need any more help.

cburgdorf commented:
@nrc Thank you very much. That was very helpful indeed. And it's a pretty cool test system I must confess!

I basically copied the existing cases and tweaked them to my needs. However, I kept some tests failing intentionally. Because I'm actually not sure what would be the expected output. For instance would we expect.

```rust
let y = Foo {
        a: x,
    };
```

```rust
let y = Foo { a: x, };
```

I would expect the former. Maybe you can look at the failing tests. All the green lines are how I would actually expect the outcome to be like. If you agree with me than I can look into tweaking the actual code to make it do the right thing ;)

nrc commented:
* the example you gave in the text, with your flag, I would expect to be formatted in the first form (unless struct_lit_style is Visual, in which case I would expect the second case, but without the comma)

* I would not expect a comma after the initialiser thing, e.g., `Foo { a, ..bar }` should not ever have a comma after `bar`

cburgdorf commented:
> I would expect to be formatted in the first form 

:+1:  me too

> unless struct_lit_style is Visual, in which case I would expect the second case, but without the comma

interesting. I'm not a fan of trailing commas (yeah, I know the cleaner git diffs etc. ;)) but I would have expected that the trailing commas folks would expect a trailing comma here even for a single first line. Anyway, I don't care ;)

>I would not expect a comma after the initialiser thin

I don't even know what this `..bar` syntax means. I've never seen this in any Rust code yet. Which probably means my Rust fu is pretty bad...

nrc commented:
I'm a fan of trailing commas, but it looks weird on one line, plus it doesn't help for diffs - the closing brace would move anyway.

The initialiser takes another struct object (of the same type) and initialises any missing fields with the fields from that struct. You don't see it very often, but it's useful in builders, etc.

cburgdorf commented:
Ok, updated. It passes all tests but I'm not sure if you like my closure and pattern matching vodoo :)

nrc commented:
Looks good, thanks for the PR!

cburgdorf commented:
Cool thanks :)
